ta/secure_key_services/src/processing_asymm.c

```diff
@@ -622,7 +622,7 @@ uint32_t do_asymm_derivation(struct pkcs11_session *session,
 	if (rv)
 		return rv;
 
-	if (get_type(*head) != SKS_CKK_GENERIC_SECRET)
+	if (get_type(*head) == SKS_CKK_GENERIC_SECRET)
```

Comment:
I think you've spot something here.
You're right: CKA_VALUE_LEN is a byte size information. So there, we must always multiply by 8 to get the bit size.

For clarity i think this should be rewritten without `key_bit_size`, here:
```patch
@@ -619,4 +619,8 @@ uint32_t do_asymm_derivation(struct pkcs11_session *session,
 	TEE_MemFill(tee_attrs, 0, sizeof(tee_attrs));
 
+	rv = get_u32_attribute(*head, SKS_CKA_VALUE_LEN, &key_byte_size);
+ 	if (rv)
+		return rv;
+
	res = TEE_AllocateTransientObject(TEE_TYPE_GENERIC_SECRET,
					  key_byte_size * 8, &out_handle);
@@ -675,6 +675,6 @@ uint32_t do_asymm_derivation(struct pkcs11_session *session,
 	if (rv)
 		goto bail;
 
-	if (a_size * 8 < key_bit_size) {
+	if (a_size < key_byte_size) {
		rv = SKS_CKR_KEY_SIZE_RANGE;
	} else {
		rv = add_attribute(head, SKS_CKA_VALUE, a_ptr, key_byte_size);
```


Note I think there is another place that may need to be fixed: in [`generate_random_key_value()`](https://github.com/foundriesio/optee-sks/blob/3f22cd7dce0dde33e58c3912364bf78aa27ef748/ta/secure_key_services/src/processing.c#L271-L291):
```patch
@@ -286,9 +286,6 @@ uint32_t do_asymm_derivation(struct pkcs11_session *session,
 		return PKCS11_CKR_ATTRIBUTE_VALUE_INVALID;
 	}
 	TEE_MemMove(&value_len, data, data_size);
 
-	if (get_type(*head) == PKCS11_CKK_GENERIC_SECRET)
-		value_len = (value_len + 7) / 8;
-
 	value = TEE_Malloc(value_len, TEE_USER_MEM_HINT_NO_FILL_ZERO);
 	if (!value)
```